docs/blueprints/selected-nodes-cache.md

```diff
@@ -0,0 +1,95 @@
+
+A node should move from the unvetted to vetted cache when:
+- when it has completed enough audits and uptime checks to be vetted
+
+## Rationale
+
+Here are some other design that were considered:
+
+1) cache the entire nodes table
+
+This approach would allow more services to use this cache and benefit from the performance gains. However, it would also add a lot of complexity. Since the nodes table is the source, we need to update the cache anytime it becomes out of sync. Some queries might be able to be stale if it doesn't impact anything else, however it might get complex keeping track of what can and can't be stale. If the entire table is cached it will become outdated very often since so many services write to the table ( see background section above for a list). A fix to this could be to have the cache be the source of truth and frequently sync the cache to the nodes table. However this would be a big undertaking since it would require changing all the services that interact with the nodes table.
+
+2) Using Redis instead of in memory cache
+
+Using a remote cache like redis, might be preferred versus in memory since this cache will be used in each satellite api replica (which can currently scale up to 12).  The more api replicas we have, the more caches if we store in memory.
+
+Concerns about using a remote cache are:
+- the network latency retrieving data from the cache. Using the cache, selecting nodes should be very fast (estimating 0.1-0.5ms), so adding a network call out to redis will make this fast cache dramtically slower.
+
+Concerns about **not** using a remote cache are:
+- that this will cause more load on the database to keep all replica caches in sync (bandwidth and load),
+- every time we update the cache we may need to lock it and potentially block other work,
+- every replica will use up this much more space in memory which will increase over time and add up with each new replica.
+
+We need to know the size this cache will be since our current redis instance handles up to 1 GB so we might need to increase the size. Currently about 80% of the nodes table qualifies to be selected for uploads, so the size of the cache for the largest cache should be about 8MB if we store the entire row in the cache, but if we only store the address and storagenode id the cache should be less than a MB.
+
+For now, lets try out using an in-memory cache for the performanace gains. If we encounter any of the concerns outlined above then we can change in the future.
+
+## Implementation
+
+Steps for implementation:
+- create a selected nodes cache in the satellite/overlay pkg
+- add redis support for this selected nodes cache
+- populate the selected nodes cache with all valida nodes from the nodes table when it is initialized
+- update the selected nodes cache when node table data changes (see list from design section above)
+- have the upload operation use the selected nodes cache
```

Comment:
If I remember correctly, I think at Space Monkey I implemented something very similar to this, but it was a pre-fetch process. So instead of waiting for an incoming upload request, it would query the node db immediately at start-up for one upload's worth of nodes, then cache them.

When an upload request came in, we would simply give them those nodes and immediately pre-fetch another batch. We had a configuration variable for how many pre-fetches we wanted stored, but eventually found that really only 1 or 2 was actually needed to keep up with incoming traffic.

This meant that (1) we never had to keep a cache up-to-date, because the pre-fetches were always _very_ recent, (2) the cache had to hold very little information, it only had list of the nodes that _would have been selected_ on the next request for a batch of nodes for upload, (3) it was crazy fast and improved perf a lot.

Just sharing as a potential thought to consider, it might work well for our use-case as well.

Reply:
I suspect that a caching the nodes with a staleness check would end up faster. The fetch from database takes quite a significant amount of data. Selecting 100 random nodes from a cached list nodes is going to be much faster. You can pick the ~100 nodes in around 600ns (+ time required to check the condition and append them into a list).

Reply:
The idea I mentioned would have it selecting from cached nodes. Essentially the difference is that rather than cache the entire table, you slowly continually fill the cache with just enough nodes to keep up with the level of requests. This way the database remains the single point of truth.

Reply:
I agree with what Egon said. All our current effort here is good, but premature with the current state.

Reply:
@isaachess I think I've been exploring an idea similar to the pre-fetch concept. I wrote up a prototype to make a queue of selected nodes here: https://github.com/storj/storj/pull/3835